src/main/scala/workflow/graph/GraphExecutor.scala

```diff
@@ -0,0 +1,81 @@
+package workflow.graph
+
+/**
+ * A GraphExecutor is constructed using a graph with already-executed state attached to it.
+ * It provides methods to further execute parts of the graph, returning the execution result.
+ * By default, it will optimize the graph before any new execution occurs.
+ *
+ * Warning: Not thread-safe!
+ *
+ * @param graph The underlying graph to execute
+ * @param optimize Whether or not to optimize the graph using the global executor before executing anything.
+ *                 Defaults to True.
+ */
+private[graph] class GraphExecutor(val graph: Graph, optimize: Boolean = true) {
+  // internally used flag to check if optimization has occurred yet. Not thread-safe!
+  private var optimized: Boolean = false
+
+  // The lazily computed result of optimizing the initial graph w/ initial state attached to it
+  private lazy val optimizedGraphAndPrefixes: (Graph, Map[NodeId, Prefix]) = {
```

Comment:
This is the comment where I realized why the prefixes map needed to be passed forward.